File: 02_CreationalDesignPatterns/02_FactoryMethod/Main.cpp
#include <iostream>
#include <string>
#include <memory>
using namespace std;

class Logistics {
public:
    virtual void send() = 0;
    virtual ~Logistics() = default;
};

class Truck : public Logistics {
public:
    void send() override {
        cout << "Delivering by Truck" << endl;
    }
};

class Ship : public Logistics {
public:
    void send() override {
        cout << "Delivering by Ship" << endl;
    }
};

class Airplane : public Logistics {
public:
    void send() override {
        cout << "Delivering by Airplane" << endl;
    }
};

class LogisticsFactory {
public:
    static unique_ptr<Logistics> createLogistics(const string& mode) {
        if (mode == "Truck") {
            return make_unique<Truck>();
        } else if (mode == "Ship") {
            return make_unique<Ship>();
        } else if (mode == "Airplane") {
            return make_unique<Airplane>();
        } else {
            throw invalid_argument("Unknown delivery mode: " + mode);
        }
    }
};

class LogisticService {
public:
    void deliver(const string& mode) {
        auto logistics = LogisticsFactory::createLogistics(mode);
        logistics->send();
    }
};

int main() {
    LogisticService service;
    service.deliver("Truck");    // Delivering by Truck
    service.deliver("Ship");     // Delivering by Ship
    service.deliver("Airplane"); // Delivering by Airplane
    // service.deliver("Bike"); // Uncommenting this line will throw an exception
    return 0;
}

File: 02_CreationalDesignPatterns/02_FactoryMethod/java/Main.java
interface Logistics {
    void send();
}

class Truck implements Logistics {
    @Override
    public void send() {
        System.out.println("Delivering by Truck");
    }
}

class Ship implements Logistics {
    @Override
    public void send() {
        System.out.println("Delivering by Ship");
    }
}

class Airplane implements Logistics {
    @Override
    public void send() {
        System.out.println("Delivering by Airplane");
    }

}

class LogisticsFactory {
    public static Logistics createLogistics(String mode) {
        switch (mode) {
            case "Truck" -> {
                return new Truck();
            }
            case "Ship" -> {
                return new Ship();
            }
            case "Airplane" -> {
                return new Airplane();
            }
            default -> throw new IllegalArgumentException("Unknown delivery mode: " + mode);
        }
    }
}

class LogisticService {
    public void deliver(String mode) {
        Logistics logistics = LogisticsFactory.createLogistics(mode);
        logistics.send();
    }
}

public class Main {
    public static void main(String[] args) {
        LogisticService service = new LogisticService();
        service.deliver("Truck"); // Delivering by Truck
        service.deliver("Ship"); // Delivering by Ship
        service.deliver("Airplane"); // Delivering by Airplane
        // service.deliver("Bike"); // Uncommenting this line will throw an exception
    }
}